keps/sig-api-machinery/20180731-crd-pruning.md

```diff
@@ -0,0 +1,610 @@
+---
+title: Pruning for Custom Resources
+authors:
+  - "@sttts"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+  - "@liggitt"
+  - "@erictune"
+  - "@mbohlool"
+  - "@apelisse"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+editor:
+  name: "@sttts"
+creation-date: 2018-07-31
+last-updated: 2019-04-30
+status: implementable
+see-also:
+ - "https://github.com/kubernetes/enhancements/pull/1002"
+ - "/keps/sig-api-machinery/20180415-crds-to-ga.md"
+---
+
+# Pruning for Custom Resources
+
+## Table of Contents
+
+   * [Pruning for Custom Resources](#pruning-for-custom-resources)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Excluding values from Pruning](#excluding-values-from-pruning)
+         * [Examples](#examples)
+         * [Opt-in and Opt-out of Pruning on CRD Level](#opt-in-and-opt-out-of-pruning-on-crd-level)
+         * [References](#references)
+         * [Test Plan](#test-plan)
+         * [Graduation Criteria](#graduation-criteria)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+      * [Version Skew Strategy](#version-skew-strategy)
+         * [Alternatives Considered](#alternatives-considered)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+CustomResources store arbitrary JSON data without following the typical Kubernetes API  behaviour to prune unknown fields. This makes CRDs different, but also leads to security and general data consistency concerns because it is unclear what is actually stored in etcd. 
+
+This KEP proposes to add pruning of all fields which are not specified in the OpenAPI validation schemas given in the CRD. 
+
+Pruning will be opt-in in `v1beta1` of `apiextensions.k8s.io` via
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+spec:
+  pruneUnknownFields: true
+  ...
+```
+
+i.e. CRDs created in `v1beta1` default to disabled pruning. 
+
+**Pruning will be enabled by default for CRDs created in `v1`**.
+
+Pruning requires _structural schemas_ (as described in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002)) for all defined versions. Validation will reject the CRD with `pruneUnknownFields: true` otherwise.
+
+Pruning can be locally disabled for subtrees of CustomResources by setting the `x-kubernetes-preserve-unknown-fields: true` vendor extension. This allows to store arbitrary JSON or RawExtensions.
```

Comment:
Is the rationale for forcing use of the openapi directive to disable pruning in `v1` to be able to surface it in published openapi? Might be worth calling that out if we actually intend to disallow disabling pruning CRD-wide in `v1`

Reply:
added a sentence that we switch from opt-in CRD-wide to opt-out for every schema.